error printed twice for deref recursion limit exceeded

I wrote some [code](https://play.rust-lang.org/?gist=1a7491f2abf052085e8b2a67e01895d6) for a rustc test which intentionally hits the recursion limit while trying to autoderef from `&Top` to `&Bottom`.

In nightly the [error when autoderef hits the recursion limit](https://github.com/rust-lang/rust/blob/aee21e2594c0dfe07f074be0da95c95b5b65ed50/src/librustc_typeck/check/autoderef.rs#L57) is printed twice. The second time it is printed without a real span.

Stable and beta:

```
rustc 1.14.0 (e8a012324 2016-12-16)
error[E0055]: reached the recursion limit while auto-dereferencing I
  --> <anon>:60:22
   |
60 |     let x: &Bottom = &t;
   |                      ^^ deref recursion limit reached

error[E0308]: mismatched types
  --> <anon>:60:22
   |
60 |     let x: &Bottom = &t;
   |                      ^^ expected struct `Bottom`, found struct `Top`
   |
   = note: expected type `&Bottom`
   = note:    found type `&Top`

error: aborting due to 2 previous errors
```

Nightly:

```
error[E0055]: reached the recursion limit while auto-dereferencing I
  --> xx.rs:60:22
   |
60 |     let x: &Bottom = &t;
   |                      ^^ deref recursion limit reached

error[E0055]: reached the recursion limit while auto-dereferencing I

error[E0308]: mismatched types
  --> xx.rs:60:22
   |
60 |     let x: &Bottom = &t;
   |                      ^^ expected struct `Bottom`, found struct `Top`
   |
   = note: expected type `&Bottom`
   = note:    found type `&Top`

error: aborting due to 3 previous errors
```

It's clearer to see with `--error-format=json` that the second error has a dummy span attached: https://gist.github.com/durka/04031f5701b5740f0f413267fc913e87

cc @jseyfried @jonathandturner

--- a/src/test/ui/issues/issue-38940.rs
+++ b/src/test/ui/issues/issue-38940.rs
@@ -1,0 +1,51 @@
+// issue-38940: error printed twice for deref recursion limit exceeded
+// Test that the recursion limit can be changed. In this case, we have
+// deeply nested types that will fail the `Send` check by overflow
+// when the recursion limit is set very low.
+#![allow(dead_code)]
+#![recursion_limit="10"]
+macro_rules! link {
+    ($outer:ident, $inner:ident) => {
+        struct $outer($inner);
+        impl $outer {
+            fn new() -> $outer {
+                $outer($inner::new())
+            }
+        }
+        impl std::ops::Deref for $outer {
+            type Target = $inner;
+            fn deref(&self) -> &$inner {
+                &self.0
+            }
+        }
+    }
+}
+struct Bottom;
+impl Bottom {
+    fn new() -> Bottom {
+        Bottom
+    }
+}
+link!(Top, A);
+link!(A, B);
+link!(B, C);
+link!(C, D);
+link!(D, E);
+link!(E, F);
+link!(F, G);
+link!(G, H);
+link!(H, I);
+link!(I, J);
+link!(J, K);
+link!(K, Bottom);
+fn main() {
+    let t = Top::new();
+    let x: &Bottom = &t;
+    //~^ ERROR mismatched types
+    //~| NOTE expected type `&Bottom`
+    //~| NOTE found type `&Top`
+    //~| NOTE expected struct `Bottom`, found struct `Top`
+    //~| ERROR reached the recursion limit while auto-dereferencing I
+    //~| NOTE deref recursion limit reached
+    //~| NOTE consider adding a `#![recursion_limit="20"]` attribute to your crate
+}

--- a/src/test/ui/issues/issue-38940.rs
+++ b/src/test/ui/issues/issue-38940.rs
@@ -42,10 +42,5 @@ fn main() {
     let t = Top::new();
     let x: &Bottom = &t;
     //~^ ERROR mismatched types
-    //~| NOTE expected type `&Bottom`
-    //~| NOTE found type `&Top`
-    //~| NOTE expected struct `Bottom`, found struct `Top`
     //~| ERROR reached the recursion limit while auto-dereferencing I
-    //~| NOTE deref recursion limit reached
-    //~| NOTE consider adding a `#![recursion_limit="20"]` attribute to your crate
 }

--- a/src/test/ui/issues/issue-38940.stderr
+++ b/src/test/ui/issues/issue-38940.stderr
@@ -1,0 +1,21 @@
+error[E0055]: reached the recursion limit while auto-dereferencing I
+  --> $DIR/issue-38940.rs:43:22
+   |
+LL |     let x: &Bottom = &t;
+   |                      ^^ deref recursion limit reached
+   |
+   = help: consider adding a `#![recursion_limit="20"]` attribute to your crate
+
+error[E0308]: mismatched types
+  --> $DIR/issue-38940.rs:43:22
+   |
+LL |     let x: &Bottom = &t;
+   |                      ^^ expected struct `Bottom`, found struct `Top`
+   |
+   = note: expected type `&Bottom`
+              found type `&Top`
+
+error: aborting due to 2 previous errors
+
+Some errors occurred: E0055, E0308.
+For more information about an error, try `rustc --explain E0055`.

